internal/cephfs/nodeserver.go

```diff
@@ -127,13 +128,47 @@ func maybeUnlockFileEncryption(
 	stagingTargetPath string,
 	volID fsutil.VolumeID,
 ) error {
-	if volOptions.IsEncrypted() {
-		log.DebugLog(ctx, "cephfs: unlocking fscrypt on volume %q path %s", volID, stagingTargetPath)
 
+	if volOptions.IsEncrypted() == false {
+		return nil
+	}
+
+	// Define Mutex Lock variables
+	lockName := string(volID) + "-mutexLock"
+	lockCookie := lockName + "-coockie"
+	lockDesc := "Lock for " + string(volID)
+	lockDuration := 40 * time.Second
+	var flags byte = 0
+
+	log.DebugLog(ctx, "Creating lock for the following volume ID %s", volID)
+
+	ioctx, err := volOptions.GetConnection().GetIoctx(volOptions.MetadataPool)
+	if err != nil {
+	}
+	defer ioctx.Destroy()
+
+	res, err := ioctx.LockExclusive(volOptions.VolID, lockName, string(lockCookie), lockDesc, lockDuration, &flags)
+	if err != nil {
+		return fmt.Errorf("Failed to lock volume ID %v: %s", volID, err)
+	}
+	log.DebugLog(ctx, "Lock successfully created for the volume ID %s", volID)
+
+	if res == 0 {
+		log.DebugLog(ctx, "cephfs: unlocking fscrypt on volume %q path %s", volID, stagingTargetPath)
 		return fscrypt.Unlock(ctx, volOptions.Encryption, stagingTargetPath, string(volID))
 	}
+	defer func() {
+		log.DebugLog(ctx, "Releasing following lock  %s", lockName)
 
-	return nil
+		_, err = ioctx.Unlock(string(volID), lockName, lockCookie)
```

Comment:
wont we get into the problem if we try to unlock if someone else is already holding the lock? in case of RWX and multiple nodes